aggregator/flush_mgr.go

```diff
@@ -0,0 +1,338 @@
+func (mgr *flushManager) flush() {
+	defer mgr.wgFlush.Done()
+
+	for {
+		var (
+			shouldFlush = false
+			duration    tally.Timer
+			flushers    []PeriodicFlusher
+			waitFor     = mgr.checkEvery
+		)
+		mgr.Lock()
+		if mgr.closed {
+			mgr.Unlock()
+			break
+		}
+		numFlushTimes := mgr.flushTimes.Len()
+		mgr.metrics.queueSize.Update(float64(numFlushTimes))
+		if numFlushTimes > 0 {
+			earliestFlush := mgr.flushTimes.Min()
+			if nowNanos := mgr.nowNanos(); nowNanos >= earliestFlush.timeNanos {
+				// NB(xichen): make a shallow copy of the flushers inside the lock
+				// and use the snapshot for flushing below.
+				shouldFlush = true
+				bucketIdx := earliestFlush.bucketIdx
+				duration = mgr.buckets[bucketIdx].duration
+				flushers = mgr.buckets[bucketIdx].flushers
+				nextFlushMetadata := flushMetadata{
+					timeNanos: earliestFlush.timeNanos + int64(mgr.buckets[bucketIdx].interval),
+					bucketIdx: bucketIdx,
+				}
+				mgr.flushTimes.Pop()
+				mgr.flushTimes.Push(nextFlushMetadata)
+			} else {
+				// NB(xichen): don't oversleep if the next flush is about to happen.
+				timeToNextFlush := time.Duration(earliestFlush.timeNanos - nowNanos)
+				if timeToNextFlush < waitFor {
+					waitFor = timeToNextFlush
+				}
+			}
+		}
+		mgr.Unlock()
+
+		if !shouldFlush {
+			mgr.sleepFn(waitFor)
+		} else {
+			var wgWorkers sync.WaitGroup
+			start := mgr.nowFn()
+			for _, flusher := range flushers {
+				flusher := flusher
+				wgWorkers.Add(1)
+				mgr.workers.Go(func() {
+					flusher.Flush()
+					wgWorkers.Done()
+				})
+			}
+			wgWorkers.Wait()
+			duration.Record(mgr.nowFn().Sub(start))
+		}
+	}
+
+	// Perform a final flush across all flushers before returning.
+	var wgWorkers sync.WaitGroup
+	for _, bucket := range mgr.buckets {
+		for _, flusher := range bucket.flushers {
+			flusher := flusher
+			wgWorkers.Add(1)
+			mgr.workers.Go(func() {
+				flusher.Flush()
+				wgWorkers.Done()
+			})
+		}
+	}
+	wgWorkers.Wait()
+}
+
+func (mgr *flushManager) nowNanos() int64 { return mgr.nowFn().UnixNano() }
+
+// flushBucket contains all the registered lists for a given flush interval.
+type flushBucket struct {
+	interval time.Duration
+	flushers []PeriodicFlusher
+	duration tally.Timer
+}
+
+func newBucket(interval time.Duration, scope tally.Scope) *flushBucket {
+	return &flushBucket{
+		interval: interval,
+		duration: scope.Timer("duration"),
+	}
+}
+
+func (b *flushBucket) Add(flusher PeriodicFlusher) {
+	b.flushers = append(b.flushers, flusher)
+}
+
+// flushMetadata contains metadata information for a flush.
+type flushMetadata struct {
+	timeNanos int64
+	bucketIdx int
+}
+
+// flushMetadataHeap is a min heap for flush metadata where the metadata with the
+// earliest flush time will be at the top of the heap. Unlike the generic heap in
+// the container/heap package, pushing data to or popping data off of the heap doesn't
+// require conversion between flush metadata and interface{}, therefore avoiding the
```

Comment:
do we need the custom impl here? the heap is not used that often and the length isn't too big?